packages/modules/product/src/services/product-module-service.ts

```diff
@@ -1339,6 +1341,13 @@ export default class ProductModuleService<
       productData.discountable = false
     }
 
+    if (productData.handle && !isValidHandle(productData.handle)) {
```

Comment:
I would create a new `validateUpdateProductInput` method, and move the validations there instead, and then we call it basically everywhere where we call `normalize*`.

Reply:
> I would create a new validateUpdateProductInput method

This makes sense. Should this be a method on the model?

> and then we call it basically everywhere where we call normalize*.

Should we validate after normalizing? Because the normalization process is supposed to create valid handles in first place.

Reply:
I believe it should be part of the service instead of the models, it create too much responsibilities for the model and if a user replace it then the validation needs to be re implemented. 

the place depends, if the handle is provided, then there might be a normalization step and it still needs to be validated right?

Reply:
> the place depends, if the handle is provided, then there might be a normalization step and it still needs to be validated right?

I have been thinking it the other way around. Lemme share my thoughts.

- If handle is provided. It should be valid and will only be validated. **Why not normalize?** Usually, its not a good idea to change the value of something a user has manually entered without any visual feedback. So if a custom handle is provided, it is only validated.
- The normalization (I think the wording is a bit off here). That is why I called it `toHandle`. This method is responsible for generate a URL-safe handle from the entity title. This method should be called when no `handle` is provided.

So in a nutshell. There is no normalization. There is

- Validation (user provided value)
- URL-safe handle generation (when no value is provided)

Reply:
@thetutlage we should validate after normalizing yes. The normalization, as Adrien said, will not create a valid handle if it is passed by the user, so the validation still makes sense.

Reply:
Alright. Jumped on a call with @sradevski and everything is clear now.

We were talking about two different functions when said `normalize`. He meant the `normalizeUpdateProductInput` or `normalizeCreateProductInput` functions and I was thinking of my `toHandle` method as normalization function.

Basically, mis-communication. We are all good and I just have to abstract the validation to its own method within the service

Reply:
no problem, sounds right 👍 , that happens sometimes especially when it is non verbal communication 😄